core/src/nucleus/ribosome/api/query.rs

```diff
@@ -59,31 +74,71 @@ pub fn invoke_query(runtime: &mut Runtime, args: &RuntimeArgs) -> ZomeApiResult
             agent.chain().query(
                 &Some(top),
                 refs.as_slice(), // Vec<&str> -> &[&str]
-                query.start,
-                query.limit,
+                ChainStoreQueryOptions {
+                    start: query.options.start,
+                    limit: query.options.limit,
+                    headers: query.options.headers,
+                }
             )
         }
     };
-    let result = match addresses {
+    let result = match maybe_result {
         // TODO #793: the Err(_code) is the RibosomeErrorCode, but we can't import that type here.
         // Perhaps return chain().query should return Some(result)/None instead, and the fixed
         // UnknownEntryType code here, rather than trying to return a specific error code.
-        Ok(result) => Ok(result),
+        Ok(result) => Ok(match (query.options.entries, result) {
+            (false, ChainStoreQueryResult::Addresses(addresses)) => {
+                QueryResult::Addresses(addresses)
+            }
+            (true,  ChainStoreQueryResult::Addresses(addresses)) => {
+                let maybe_entries: Result<Vec<Entry>,HolochainError> = addresses
+                    .iter()
+                    .map(|address| // -> Result<Entry, HolochainError>
+                         Ok(
+                             get_entry_from_dht(&runtime.context, address.to_owned())? // -> Result<Option<Entry>, HolochainError>
+                                 .ok_or(HolochainError::ErrorGeneric(
+                                     format!("Failed to obtain Entry for Address {}", address)))?))
+                    .collect();
+
+                match maybe_entries {
+                    Ok(entries) => QueryResult::Entries(entries),
+                    Err(_e) => return ribosome_error_code!(UnknownEntryType), // TODO: return actual error?
+                }
+            }
+            (false, ChainStoreQueryResult::Headers(headers)) => QueryResult::Headers(headers),
+            (true,  ChainStoreQueryResult::Headers(headers)) => {
+                let maybe_headers_with_entries: Result<Vec<(ChainHeader,Entry)>,HolochainError> = headers
+                    .iter()
+                    .map(|header| // -> Result<Entry, HolochainError>
+                         Ok((header.to_owned(),
+                             get_entry_from_dht(&runtime.context, header.entry_address().to_owned())? // -> Result<Option<Entry>, HolochainError>
+                             .ok_or(HolochainError::ErrorGeneric(
+                                 format!("Failed to obtain Entry for Address {}", header.entry_address())))?)))
+                    .collect();
+                match maybe_headers_with_entries {
+                    Ok(headers_with_entries) => QueryResult::HeadersWithEntries(headers_with_entries),
+                    Err(_e) => return ribosome_error_code!(UnknownEntryType), // TODO: return actual error?
```

Comment:
If it works, this is probably fine now, but would rather see the entry-getting stuff DRYed up and refactored into a function